README.md
# livelex-qt
Small Python library to use livelex with Qt's QTextDocument

This module depends on livelex.

Homepage: https://github.com/wbsoft/livelex-qt


Example:

```python

from PyQt5.QtWidgets import QApplication
from PyQt5.QtGui import QTextDocument

app = QApplication([])

import livelexqt
import livelex.lang.xml

lexicon = livelex.lang.xml.Xml.root
doc = QTextDocument()
livelexqt.SyntaxHighlighter.instance(doc, lexicon)

# Now the text in the document is automatically highlighted using the
# specified root lexicon.

```


ChangeLog
ChangeLog
=========


2023-05-28: parceqt-0.33.0

- fixes a TypeError with recent Python and recent PyQt5: supply int to QColor()


2022-03-14: parceqt-0.32.0

- requires parce 0.32.0
- added gadgets module with ExtraSelectionManager and LanguageMenuAction
- added widgets module with LexiconChooser
- debugger: added language selection in View menu
- simplified handling of singleton QObject instances.


2022-02-12: parceqt-0.31.0

- requires parce 0.31.0
- Document() now also can create QTextDocument; use Document(doc) to wrap an
  existing document
- small adaptations to parce API changes


2021-12-09: parceqt-0.27.0

- requires parce 0.27
- add work module and adapt design to new parce Document/Worker
- add Guess button to debugger, to guess the language


2021-11-11: parceqt-0.25.0

- bring version number and some API changes on par with parce
- adjust debugger to API changes of parce.language; now use parce.registry


2021-01-21: parceqt-0.17.0

- added Cursor.html() and copy_html() convenience methods
- moved debugging logic (signals etc) to main treebuilder
- debug: removed DebugFormatter, parce can now handle debug formatting itself
- debug window: larger text font by default
- fixed black colors in palette when textformat didn't specify color
- fixed other little bugs
- small api adjustments to parce-0.17.0


2021-01-17: parceqt-0.16.0

- added Edit → Copy selection as HTML
- parceqt.cursor(): fixed new attrname cursor.start → cursor.pos


2021-01-12: parceqt-0.15.0

- small api adjustments to parce-0.15.0
- some cleanups
- documentation switched to Furo theme, looks better on small devices


2021-01-02: parceqt-0.14.0

- added Document.find_start/end_of_line() reimplementation using QTextBlock
- small api adjustments to parce-0.14.0


2020-06-02: parceqt-0.13.0

- small api adjustments to parce-0.13.0


2020-05-18: parceqt-0.12.0

- very small api adjustments to parce-0.12.0


2020-05-13: parceqt-0.10.0

- very small api adjustments to parce-0.10.0


2020-05-04: parceqt-0.9.0

- some small api adjustments to parce-0.9.0
- set window title of debugger window


2020-04-25: parceqt-0.8.5

- add "Debug Unparsed Text" theme, highlighting unparsed text
- adjust treebuilder and highlighter to the new peek() call of parce.treebuilder
- debug window:
  - when opening a file, automatically guess the language to use
  - new commands to show/hide the tree and to reload the theme stylesheet
  - Ctrl-wheel zoom now changes the font size of the text editor window
  - add expand/collapse actions
  - ability to run the debug window with python3 -m parceqt.debug <filename>


2020-04-07: parceqt-0.8.0

- added treemodel.TreeModel
- added debug module with debugger window
- use new peek feature to quickly update highlighting surrounding change


2020-03-14: parceqt-0.7.0

- don't set a default Theme on init of SyntaxHighlighter, this causes
  the highlighting to run twice when calling the highlight() helper.
- really wait til update on Treebuilder.wait()


2020-02-16: parceqt-0.6.0

- Improvements, and adjust to changes in parceqt-0.6
- Convenience function to adjust a text widget to the window() format
  of a theme.


2020-02-07: parceqt-0.5.0

- API to use highlighter and set root lexicon greatly simplfied


2020-01-22: parceqt-0.3.0

- Renamed to parceqt


2020-01-21: livelex-qt-0.3.0

- Better documentation
- Split classes over separate modules files, accessible via the toplevel
  package
- robustness improvements, highlighting must happen in the main thread but
  it regularly polls for user events and restarts if needed.


2020-01-15: livelex-qt-0.2.0

- Incorporate qtdocument module from livelex
- Initial release



parceqt/formatter.py
# -*- coding: utf-8 -*-
#
# This file is part of the parce-qt Python package.
#
# Copyright © 2020 by Wilbert Berendsen <info@wilbertberendsen.nl>
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This module is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


"""
A Formatter that uses parce.theme.Theme with QTextCharFormat.
"""


from PyQt5.QtGui import QColor, QFont, QPalette, QTextCharFormat
from PyQt5.QtWidgets import QApplication

import parce.formatter


class Formatter(parce.formatter.Formatter):
    """Formatter, inheriting from parce.Formatter, but using Qt text formats by default."""
    def __init__(self, theme, factory=None):
        """Reimplemented to use the text_format factory by default."""
        super().__init__(theme, factory or text_format)

    def font(self, widget=None):
        """Return the font of the default text format.

        This font can then be used for a text editing widget. If widget is
        specified, resolves the application's default font for the widget with
        the properties from the theme.

        """
        font = QApplication.font(widget)
        f = self.baseformat()
        if f:
            font = f.font().resolve(font)
        return font

    def palette(self, widget=None):
        """Return a QPalette with the following colors set:

        ``QPalette.Text``
            default foreground color
        ``QPalette.Base``
            default background color
        ``QPalette.HighlightText``
            selection foreground color
        ``QPalette.Highlight``
            selection background color
        ``QPalette.AlternateBase``
            background color for the current line

        If the theme supports it, the Inactive and Disabled color groups are
        set to their own colors. Otherwise, they just use the same colors.

        If widget is specified, resolves the application's default palette for
        the widget with the properties from the theme.

        """
        # all color groups
        p = QApplication.palette(widget)
        f = self.baseformat("window")  # QTextCharFormat
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.Text, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Base, f.background().color())
        f = self.baseformat("selection")
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.HighlightedText, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Highlight, f.background().color())
        f = self.baseformat("current-line")
        if f and f.hasProperty(QTextCharFormat.BackgroundBrush):
            p.setColor(QPalette.AlternateBase, f.background().color())

        # Active color group
        f = self.baseformat("window", "focus")
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.Active, QPalette.Text, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Active, QPalette.Base, f.background().color())
        f = self.baseformat("selection", "focus")
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.Active, QPalette.HighlightedText, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Active, QPalette.Highlight, f.background().color())
        f = self.baseformat("current-line", "focus")
        if f and f.hasProperty(QTextCharFormat.BackgroundBrush):
            p.setColor(QPalette.Active, QPalette.AlternateBase, f.background().color())

        # Disabled color group
        f = self.baseformat("window", "disabled")
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.Disabled, QPalette.Text, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Disabled, QPalette.Base, f.background().color())
        f = self.baseformat("selection", "disabled")
        if f:
            if f.hasProperty(QTextCharFormat.ForegroundBrush):
                p.setColor(QPalette.Disabled, QPalette.HighlightedText, f.foreground().color())
            if f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Disabled, QPalette.Highlight, f.background().color())
        f = self.baseformat("current-line", "disabled")
        if f and f.hasProperty(QTextCharFormat.BackgroundBrush):
                p.setColor(QPalette.Disabled, QPalette.AlternateBase, f.background().color())
        return p


def _color(c):
    """Convert css.Color to QColor."""
    return QColor(c.r, c.g, c.b, int(c.a * 255))


def _font_point_size(size, unit):
    """Return a suitable point size, where 12 is a default value."""
    if isinstance(size, str):
        return {
            "xx-small": 8,
            "x-small": 9,
            "small": 10,
            "medium": 12,
            "large": 14,
            "x-large": 16,
            "xx-large": 20,
            "xxx-large": 24,
            "larger": 14,
            "smaller": 10,
        }.get(size, 12)
    elif unit == "pt":
        return size
    elif unit == "px":
        return size * 12 / 16
    elif unit in ("rem", "em"):
        return size * 12
    elif unit == "%":
        return size * 12 / 100
    return 12


def _font_stretch(stretch):
    """Return a suitable font stretch."""
    if isinstance(stretch, str):
        return {
            "ultra-condensed": 50,
            "extra-condensed": 62,
            "condensed": 75,
            "semi-condensed": 87,
            "normal": 100,
            "semi-expanded": 112,
            "expanded": 125,
            "extra-expanded": 150,
            "ultra-expanded": 200,
        }.get(stretch, 100)
    return int(stretch)


def _font_weight(weight):
    """Return a suitable weight."""
    if isinstance(weight, str):
        if weight == "bold":
            return QFont.Bold
        elif weight == "bolder":
            return QFont.ExtraBold
        elif weight == "lighter":
            return QFont.Light
    else:
        if weight >= 900:
            return QFont.Black
        elif weight >= 800:
            return QFont.ExtraBold
        elif weight >= 700:
            return QFont.Bold
        elif weight >= 600:
            return QFont.DemiBold
        elif weight >= 500:
            return QFont.Medium
        elif weight >= 400:
            return QFont.Normal
        elif weight >= 300:
            return QFont.Light
        elif weight >= 200:
            return QFont.ExtraLight
        else:
            return QFont.Thin
    return QFont.Normal


def text_format(tf):
    """A factory to be used with parce.theme.Formatter.

    Creates a QTextCharFormat for the specified TextFormat object.

    """
    if tf:
        f = QTextCharFormat()
        if tf.color:
            f.setForeground(_color(tf.color))
        if tf.background_color:
            f.setBackground(_color(tf.background_color))
        if tf.text_decoration_line:
            if 'underline' in tf.text_decoration_line:
                f.setFontUnderline(True)
            if 'overline' in tf.text_decoration_line:
                f.setFontOverline(True)
            if 'line-through' in tf.text_decoration_line:
                f.setFontStrikeOut(True)
        if tf.text_decoration_style:
            s = tf.text_decoration_style
            if s == "solid":
                f.setUnderlineStyle(QTextCharFormat.SingleUnderline)
            #elif s == "double":
            #    pass # Seems Qt5 does not provide this
            elif s == "dotted":
                f.setUnderlineStyle(QTextCharFormat.DotLine)
            elif s == "dashed":
                f.setUnderlineStyle(QTextCharFormat.DashUnderline)
            elif s == "wavy":
                f.setUnderlineStyle(QTextCharFormat.WaveUnderline)
        if tf.text_decoration_color:
            f.setUnderlineColor(_color(tf.text_decoration_color))
        if tf.font_family:
            try:
                f.setFontFamilies(tf.font_family)
            except AttributeError: # this property was introduced in Qt 5.13
                pass
            f.setFontFamily(tf.font_family[0])
        if tf.font_size:
            f.setFontPointSize(_font_point_size(tf.font_size, tf.font_size_unit))
        if tf.font_stretch:
            f.setFontStretch(_font_stretch(tf.font_stretch))
        if tf.font_style in ('italic', 'oblique'):
            f.setFontItalic(True)
        if tf.font_variant_caps == "small-caps":
            f.setFontCapitalization(QFont.SmallCaps)
        if tf.font_weight:
            f.setFontWeight(_font_weight(tf.font_weight))
        if not f.isEmpty():
            return f

